src/ert/storage/migration/to9.py

```diff
@@ -0,0 +1,87 @@
+    """
+    Writes the data to the filename as a transaction.
+
+    Guarantees to not leave half-written or empty files on disk if the write
+    fails or the process is killed.
+    """
+
+    Path("./swp").mkdir(parents=True, exist_ok=True)
+    with NamedTemporaryFile(dir="./swp", delete=False) as f:
+        f.write(data)
+        os.rename(f.name, filename)
+
+
+def migrate(path: Path) -> None:
+    for experiment in path.glob("experiments/*"):
+        ensembles = path.glob("ensembles/*")
+
+        with (
+            open(experiment / "index.json", encoding="utf-8") as f_experiment,
+            open(
+                experiment / "responses.json", mode="r", encoding="utf-8"
+            ) as f_responses,
+        ):
+            exp_index = json.load(f_experiment)
+            experiment_id = exp_index["id"]
+
+            responses_config = json.load(f_responses)
+            for response_type, config in responses_config.items():
+                if "has_finalized_keys" not in config:
+                    # Read a sample response and write the keys
+                    for ens in ensembles:
+                        with open(ens / "index.json", encoding="utf-8") as f_ensemble:
+                            ens_file = json.load(f_ensemble)
+                            if ens_file["experiment_id"] != experiment_id:
+                                continue
+
+                        real_dirs = [*ens.glob("realization-*")]
+
+                        for real_dir in real_dirs:
+                            if (real_dir / f"{response_type}.parquet").exists():
+                                df = polars.scan_parquet(
+                                    real_dir / f"{response_type}.parquet"
+                                )
+                                response_keys = (
+                                    df.select("response_key")
+                                    .unique()
+                                    .collect()
+                                    .to_series()
+                                    .to_list()
+                                )
+                                config["has_finalized_keys"] = True
+                                config["keys"] = sorted(response_keys)
+                                break
+
+                        if config["has_finalized_keys"]:
+                            break
+
+                # If this is hit, it means no responses were found
+                # for that response type, so we still cannot have "finalized"
+                # keys. We then default it to that of the configs.
+                # At time of writing, it is always True for GenDataConfig
+                # and False for SummaryConfig
+                if "has_finalized_keys" not in config:
+                    config["has_finalized_keys"] = (
+                        config["_ert_kind"] != "SummaryConfig"
+                    )
+
+        os.remove(experiment / "responses.json")
```

Comment:
Then this should be removed?

Reply:
If the transaction just overwrites yes? AFAIK it just writes (maybe that is also an overwrite). Why not keep the removal explicit though? I'm not sure if the transaction is meant to captured file writing failures only, or also file overwriting failures and preserve the to-be-overwritten file?

Reply:
It was if the transaction fail, if it would it would be left in an inconsistent state

Reply:
Did not understand at first, but think I see now, if we don't remove it we'll be able to retry doing the migration in this case